nav2_graceful_motion_controller/src/graceful_motion_controller.cpp

```diff
@@ -0,0 +1,344 @@
+  }
+
+  return *goal_pose_it;
+}
+
+bool GracefulMotionController::simulateTrajectory(
+  const geometry_msgs::msg::PoseStamped & robot_pose,
+  const geometry_msgs::msg::PoseStamped & motion_target, nav_msgs::msg::Path & trajectory,
+  const bool & backward)
+{
+  // Check for collision before moving
+  if (inCollision(
+      robot_pose.pose.position.x, robot_pose.pose.position.y,
+      tf2::getYaw(robot_pose.pose.orientation)))
+  {
+    return false;
+  }
+
+  // First pose
+  geometry_msgs::msg::PoseStamped next_pose;
+  next_pose.header.frame_id = costmap_ros_->getBaseFrameID();
+  next_pose.pose.orientation.w = 1.0;
+  trajectory.poses.push_back(next_pose);
+
+  double distance = std::numeric_limits<double>::max();
+  double resolution_ = costmap_ros_->getCostmap()->getResolution();
+  double dt = (params_->v_linear_max > 0.0) ? resolution_ / params_->v_linear_max : 0.0;
+
+  // Set max iter to avoid infinite loop
+  unsigned int max_iter = 2 * sqrt(
+    motion_target.pose.position.x * motion_target.pose.position.x +
+    motion_target.pose.position.y * motion_target.pose.position.y) / resolution_;
+
+  // Generate path
+  do{
+    // Apply velocities to calculate next pose
+    next_pose.pose = control_law_->calculateNextPose(
+      dt, motion_target.pose, next_pose.pose, backward);
+
+    // Add the pose to the trajectory for visualization
+    trajectory.poses.push_back(next_pose);
+
+    // Check for collision
+    geometry_msgs::msg::PoseStamped global_pose;
+    tf2::doTransform(next_pose, global_pose, costmap_transform_);
+    if (inCollision(
+        global_pose.pose.position.x, global_pose.pose.position.y,
+        tf2::getYaw(global_pose.pose.orientation)))
+    {
+      return false;
+    }
+
+    // Check if we reach the goal
+    distance = nav2_util::geometry_utils::euclidean_distance(motion_target.pose, next_pose.pose);
+  }while(distance > resolution_ && trajectory.poses.size() < max_iter);
+
+  return true;
+}
+
+geometry_msgs::msg::Twist GracefulMotionController::rotateToTarget(const double & angle_to_target)
+{
+  geometry_msgs::msg::Twist vel;
+  vel.linear.x = 0.0;
+  vel.angular.z = params_->rotation_scaling_factor * angle_to_target * params_->v_angular_max;
+  return vel;
+}
+
+bool GracefulMotionController::inCollision(const double & x, const double & y, const double & theta)
+{
+  unsigned int mx, my;
+  if (!costmap_ros_->getCostmap()->worldToMap(x, y, mx, my)) {
+    RCLCPP_WARN(
+      logger_, "The path is not in the costmap. Cannot check for collisions. "
+      "Proceed at your own risk, slow the robot, or increase your costmap size.");
+    return false;
+  }
+
+  // Calculate the cost of the footprint at the robot's current position depending
+  // on the shape of the footprint
+  double footprint_cost;
+  if (costmap_ros_->getUseRadius()) {
+    footprint_cost = collision_checker_->pointCost(mx, my);
+    if (footprint_cost >= nav2_costmap_2d::INSCRIBED_INFLATED_OBSTACLE) {
```

Comment:
I think this misses the condition of unknown space like handled in the full footprint case.

I think in general, this logic can be lifted from somewhere else in the stack to make sure it handles all the conditions. I'm only glancing at this quickly, but I don't think this is the only problem

Reply:
Mmm, I think I don't understand this. Isn't unknown space the same as NO_INFORMATION = 255?

With that logic, any value above 253 will be treated as a collision.

Reply:
Still a problem, I think